File: lista2/java/FiniteField.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class FiniteField extends Field {
    private int value;
    private static final int characteristic = 1234567891;
    public FiniteField() {
        this.value = 0;
    }

    public FiniteField(int value) {
        setValue(value);
    }

    public void setValue(int value) {
        while(value < 0) {
            value += characteristic;
        }

        this.value = value % characteristic;
    }

    public int getValue() {
        return this.value;
    }

    public int getCharacteristic() {
        return characteristic;
    }

    private FiniteField findInverse(int value) {
        int[] res = new int[2];
        try {
            int g = gcdExtended(value, characteristic, res);
            if (g != 1) {
                throw new Exception();
            }
            else {
                return new FiniteField(res[0]);
            }
        }
        catch(Exception e)
        {
            System.out.println("Inverse not found");
        }

        return new FiniteField();
    }

    private int gcdExtended(int a, int b, int[] res) {
        if (a == 0) {
            res[0] = 0;
            res[1] = 1;
            return b;
        }

        // To store results of recursive call
        int gcd = gcdExtended(b % a, a, res);
        int x1 = res[0];
        int y1 = res[1];

        // Update x and y using results of recursive
        // call
        int tmp = b / a;
        res[0] = y1 - tmp * x1;
        res[1] = x1;

        return gcd;
    }

    public Field add(final Field obj) {
        return new FiniteField(this.value + obj.getValue());
    }

    public Field sub(final Field obj) {
        int result = this.value - obj.getValue();
        return new FiniteField(result);
    }

    public Field mult(final Field obj) {
        int res = fastMultiplication(this.value, obj.getValue());
        return new FiniteField(res);
    }

    private int fastMultiplication(int a, int b) {
        int res = 0;

        a %= characteristic;
        while (b > 0) {

            if ((b & 1) > 0) {
                res = (res + a) % characteristic;
            }
            a = (2 * a) % characteristic;
            b >>= 1;
        }
        return res;
    }

    public Field div(final Field obj) {
        FiniteField inv = findInverse(obj.getValue());
        return mult(inv);
    }

    public void assign(final Field obj) {
        this.value = obj.getValue();
    }

    public void multAndAssign(final Field obj) {
        assign(mult(obj));
    }

    public void divAndAssign(final Field obj) {
        assign(div(obj));
    }

    public void addAndAssign(final Field obj) {
        assign(add(obj));
    }

    public void subAndAssign(final Field obj) {
        assign(sub(obj));
    }

    public boolean ifNotEqual(final Field b) {
        return this.value != b.getValue();
    }

    public boolean ifEqual(final Field b) {
        return this.value == b.getValue();
    }

    public boolean ifLessEqual(final Field b) {
        return this.value <= b.getValue();
    }

    public boolean ifGreaterEqual(final Field b) {
        return this.value >= b.getValue();
    }

    public boolean ifLess(final Field b) {
        return this.value < b.getValue();
    }

    public boolean ifGreater(final Field b) {
        return this.value > b.getValue();
    }

    public static void stdinRead(Field a)  {
        try {
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            FiniteField b = new FiniteField(Integer.parseInt(br.readLine()));
            a.assign(b);
        } catch (IOException ex){
            System.out.println("IO exception");
        }
    }
    @Override
    public String toString() {
        return String.valueOf(value);
    }
}

File: lista2/python/FiniteField.py
class gfException(Exception):
    pass

class FiniteField:
    characteristic = 1234577
    
    def __init__(self, value=0):
        self.setValue(value)

    def getValue(self):
        return self.value

    def setValue(self, value):
        while value < 0:
            value += FiniteField.characteristic
        
        self.value = value % FiniteField.characteristic

    @staticmethod
    def getCharacteristic():
        return FiniteField.characteristic

    @staticmethod
    def gcdExtended(a, b):
        if a == 0:
            return b, 0, 1
        gcd, x1, y1 = FiniteField.gcdExtended(b % a, a)
        x = y1 - (b // a) * x1
        y = x1
        return gcd, x, y

    @staticmethod
    def findInverse(value):
        try:
            g, x, y = FiniteField.gcdExtended(value, FiniteField.characteristic)
            if g != 1:
                raise gfException("GF field exception")
            else:
                return FiniteField((x % FiniteField.characteristic) % FiniteField.characteristic)
        except gfException as e:
            print(e)

        return FiniteField()

    def assign(self, obj):
        self.value = obj.value
        return self

    #use >> in c style
    def __rshift__(self, in_stream):
        try:
            value = int(in_stream.readline().strip())
            self.setValue(value)
        except Exception as e:
            print(e)
        return self

    #print(object)
    def __str__(self):
        return str(self.value)
    
    def __add__(self, obj):
        return FiniteField((self.value + obj.value) % FiniteField.characteristic)

    def __sub__(self, obj):
        result = self.value - obj.value
        return FiniteField(characteristic + result) if result < 0 else FiniteField(result)

    def __mul__(self, obj):
        res = FiniteField(0)
        for _ in range(obj.value):
            res += self
        return res

    def __truediv__(self, obj):
        res = self.findInverse(obj.value)
        return self * res

    def __iadd__(self, obj):
        self.setValue((self.value + obj.value) % FiniteField.characteristic)
        return self

    def __isub__(self, obj):
        result = self.value - obj.value
        self.setValue(FiniteField.characteristic + result if result < 0 else result)
        return self

    def __imul__(self, obj):
        temp = obj.value
        res = FiniteField(0)
        for _ in range(temp):
            res += self
        self.setValue(res.value)
        return self

    def __itruediv__(self, obj):
        res = self.findInverse(obj.value)
        self *= res
        return self

    def __eq__(self, obj):
        return self.value == obj.value

    def __ne__(self, obj):
        return not self == obj

    def __le__(self, obj):
        return self.value <= obj.value

    def __ge__(self, obj):
        return self.value >= obj.value

    def __lt__(self, obj):
        return self.value < obj.value

    def __gt__(self, obj):
        return self.value > obj.value